lib/ide/ide/src/view/view_layout.rs

```diff
@@ -0,0 +1,151 @@
+        self.view_mode = view_mode;
+        self.update();
+    }
+
+    fn switch_mode(&mut self) {
+        if let ViewMode::Half = self.view_mode {
+            self.set_view_mode(ViewMode::Full)
+        } else {
+            self.set_view_mode(ViewMode::Half)
+        }
+    }
+
+    fn set_dimensions(&mut self, dimensions:Vector2<f32>) {
+        self.dimensions = dimensions;
+        self.update();
+    }
+
+    fn update(&mut self) {
+        let dimensions = self.dimensions;
+        let (position,dimensions) = match self.view_mode {
+            ViewMode::Full => {
+                let position   = Vector2::new(0.0, dimensions.y);
+                (position,dimensions)
+            },
+            ViewMode::Half => {
+                let position   = Vector2::new(0.0, dimensions.y / 2.0);
+                let dimensions = Vector2::new(dimensions.x, dimensions.y / 2.0);
+                (position,dimensions)
+            }
+        };
+        self.text_editor.set_dimensions(dimensions);
+        self.text_editor.set_position(position);
+        self.text_editor.update();
+    }
+}
+
+
+
+// ==================
+// === ViewLayout ===
+// ==================
+
+/// Initial implementation of ViewLayout with a single TextEditor. Pressing ctrl+f toggles
+/// fullscreen mode.
+#[derive(Debug,Clone)]
+pub struct ViewLayout {
+    data : Rc<RefCell<ViewLayoutData>>
+}
+
+impl ViewLayout {
+    /// Creates a new ViewLayout with a single TextEditor.
+    pub fn default(world:&World) -> Self {
+        let text_editor      = TextEditor::new(&world);
+        let keyboard_closure = None;
+        let view_mode        = ViewMode::Half;
+        let dimensions       = zero();
+        let data             = ViewLayoutData{text_editor,keyboard_closure,view_mode,dimensions};
+        let data             = Rc::new(RefCell::new(data));
+        Self {data}.init(world)
+    }
+
+    fn init_keyboard(self) -> Self {
+        let data    = Rc::downgrade(&self.data);
+        let closure = move |event:KeyboardEvent| {
+            const F_KEY : u32 = 70;
+            if event.ctrl_key() && event.key_code() == F_KEY {
+                data.upgrade().map(|data| {
+                    data.borrow_mut().switch_mode()
+                });
+                event.prevent_default();
+            }
+        };
+        let closure : Box<dyn FnMut(KeyboardEvent)> = Box::new(closure);
+        let keyboard_closure                        = Closure::wrap(closure);
+        let callback : &Function                    = keyboard_closure.as_ref().unchecked_ref();
+        let body = document().unwrap().body().unwrap();
+        body.add_event_listener_with_callback("keydown", callback).ok();
+        self.data.borrow_mut().keyboard_closure = Some(keyboard_closure);
+        self
+    }
+
+    fn init(mut self, world:&World) -> Self {
+        let screen = world.scene().camera().screen();
+        let dimensions = Vector2::new(screen.width,screen.height);
```

Comment:
Why camera uses size types incompatible with this here?
We should have a dictionary type like `Size2<f32>` and have all libraries use it on API boundary. It seems strange, that view_layout needs to know about camera's implementation details (like screen field names), just to rewrap it in a different structure used elsewhere.

Reply:
About your concern with dimensions and size. Yes, there is this disparity in the API regarding naming things as dimensions or size. I can rework that. About defining a newtype Size2<f32>, some time ago it was defined that we will make direct use of Vector2<f32>, but I am open to change it. I will standardize the interfaces anyway.